src/gui/guiScene.cpp

```diff
@@ -0,0 +1,239 @@
+		return m_mesh;
+	} else if (m_mesh) {
+		m_mesh->remove();
+		m_mesh = nullptr;
+	}
+
+	m_mesh = m_smgr->addAnimatedMeshSceneNode(mesh);
+	m_mesh->setPosition(-m_mesh->getBoundingBox().getCenter());
+	m_mesh->animateJoints();
+
+	return m_mesh;
+}
+
+void GUIScene::setTexture(u32 idx, video::ITexture *texture)
+{
+	video::SMaterial &material = m_mesh->getMaterial(idx);
+	material.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
+	material.MaterialTypeParam = 0.5f;
+	material.TextureLayer[0].Texture = texture;
+	material.setFlag(video::EMF_LIGHTING, false);
+	material.setFlag(video::EMF_FOG_ENABLE, true);
+	material.setFlag(video::EMF_BILINEAR_FILTER, false);
+	material.setFlag(video::EMF_BACK_FACE_CULLING, false);
+}
+
+void GUIScene::draw()
+{
+	core::rect<s32> oldViewPort = m_driver->getViewPort();
+	m_driver->setViewPort(getAbsoluteClippingRect());
+	core::recti borderRect = Environment->getRootGUIElement()->getAbsoluteClippingRect();
+
+	if (m_bgcolor != 0) {
+		Environment->getSkin()->draw3DSunkenPane(
+			this, m_bgcolor, false, true, borderRect, 0);
+	}
+
+	core::dimension2d<s32> size = getAbsoluteClippingRect().getSize();
+	m_smgr->getActiveCamera()->setAspectRatio((f32)size.Width / (f32)size.Height);
+
+	if (!m_target) {
+		updateCamera(m_smgr->addEmptySceneNode());
+		rotateCamera(v3f(0.f));
+		m_cam->bindTargetAndRotation(true);
+	}
+
+	cameraLoop();
+
+	if (m_autorotate)
+		rotateCamera(v3f(0.f, -0.5f, 0.f));
+
+	m_smgr->drawAll();
+
+	if (m_mesh) {
+		calcOptimalDistance();
+		m_mesh = nullptr;
+	}
+
+	m_driver->setViewPort(oldViewPort);
+}
+
+bool GUIScene::OnEvent(const SEvent &event)
+{
+	if (m_mouse_ctrl) {
+		switch(event.EventType) {
+		case EET_MOUSE_INPUT_EVENT:
+			if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+				lastPos = v2f((f32)event.MouseInput.X, (f32)event.MouseInput.Y);
+				return true;
+			} else if (event.MouseInput.Event == EMIE_MOUSE_MOVED) {
+				if (event.MouseInput.isLeftPressed()) {
+					currPos = v2f((f32)event.MouseInput.X, (f32)event.MouseInput.Y);
+					rotateCamera(v3f(
+						lastPos.Y - currPos.Y, currPos.X - lastPos.X, 0.f));
+					lastPos = currPos;
+					return true;
+				}
+			}
+			break;
+		default: break;
+		}
+	}
+
+	return gui::IGUIElement::OnEvent(event);
+}
+
+/* Camera control functions */
+
+inline void GUIScene::calcOptimalDistance()
+{
+	core::aabbox3df box = m_mesh->getBoundingBox();
+	f32 width  = box.MaxEdge.X - box.MinEdge.X;
+	f32 height = box.MaxEdge.Y - box.MinEdge.Y;
+	f32 depth  = box.MaxEdge.Z - box.MinEdge.Z;
+	f32 max_width = width > depth ? width : depth;
+
+	const scene::SViewFrustum *f = m_cam->getViewFrustum();
+	f32 far = m_cam->getFarValue();
+	f32 far_width = core::line3df(f->getFarLeftUp(), f->getFarRightUp()).getLength();
+	f32 far_height = core::line3df(f->getFarLeftUp(), f->getFarLeftDown()).getLength();
+
+	core::recti rect = getAbsolutePosition();
+	f32 zoomX = rect.getWidth() / max_width;
+	f32 zoomY = rect.getHeight() / height;
+	f32 dist;
```

Comment:
Works well enough already. I'd like to point out an edge-case where the chest chopped at the bottom by a little bit: 

![grafik](https://user-images.githubusercontent.com/1497498/90286270-c708bd00-de75-11ea-9232-4c5efb3dfbcb.png)

Reply:
It calculates the optimal distance when the mesh is in its initial position. Here the mesh has moved.

Reply:
That explains it very well. Thank you.